include/thread_queue.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include "async_event.h"
+#include "double_buffered_queue.h"
+#include "named_threads.h"
+#include "utils.h"
+
+#include <chrono>
+#include <future>
+#include <thread>
+#include <type_traits>
+#include <variant>
+
+namespace celerity::detail {
+
+/// A single-thread job queue accepting functors and returning events that conditionally forward job results.
+class thread_queue {
+  public:
+	/// Constructs a null thread queue that cannot receive jobs.
+	thread_queue() = default;
+
+	/// Spawns a new thread queue with the given thread name. If `enable_profiling` is set to `true`, completed events from this thread queue will report a
+	/// non-nullopt duration.
+	explicit thread_queue(std::string thread_name, const bool enable_profiling = false) : m_impl(new impl(std::move(thread_name), enable_profiling)) {}
+
+	// thread_queue is movable, but not copyable.
+	thread_queue(const thread_queue&) = delete;
+	thread_queue(thread_queue&&) = default;
+	thread_queue& operator=(const thread_queue&) = delete;
+	thread_queue& operator=(thread_queue&&) = default;
+
+	/// Destruction will await all submitted and pending jobs.
+	~thread_queue() {
+		if(m_impl != nullptr) {
+			m_impl->queue.push(job{} /* termination */);
+			m_impl->thread.join();
+		}
+	}
+
+	/// Submit a job to the thread queue.
+	/// `fn` must take no arguments and return either `void` or a type convertible to `void *`, which will be forwarded as the result into the event.
+	template <typename Fn>
+	async_event submit(Fn&& fn) {
```

Comment:
```suggestion
	async_event submit(Fn&& fn) {
            static_assert(std::is_same_v<std::invoke_result_t<Fn>, void> || std::is_convertible_v<std::invoke_result_t<Fn>, void*>);
```
?

Reply:
Sure! Also added some `std::decay_t` and `mutable`s while I was at it.